### PR TITLE
Go.c to d

### PR DESCRIPTION
**Pull request by WalterBright**

### CODE REVIEW

File: src/dmd/backend/go.d [hunk 1]
```diff
@@ -0,0 +1,500 @@
...
+
+extern (C++) void mem_free(void* p);
+
+void go_term()
+{
+    vec_free(go.defkill);
+    vec_free(go.starkill);
+    vec_free(go.vptrkill);
+version (Posix)
+{
```
Comment (jacob-carlborg) [inline src/dmd/backend/go.d]: Seems to be some unusual indentations in some places, like this one.

Comment (WalterBright) [reply src/dmd/backend/go.d]: I was trying to minimize diffs with the C version.

Comment (jacob-carlborg) [reply src/dmd/backend/go.d]: I suspected that.

Comment (jacob-carlborg) [reply src/dmd/backend/go.d]: Ah, looking at the individual commits worked a lot better, now it makes more sense. 👍

**Comment by RazvanN7** [comment]
Are you going to make any measurements  to see how the conversion to D affects the compilation time?

**Comment by WalterBright** [comment]
Don't bother reviewing this till I figure out what is going wrong with the OSX32 build.

**Comment by WalterBright** [comment]
Hmm, looks like `clock()` on OSX32 is the source of the problem.

**Comment by WalterBright** [comment]
Definitely it's clock():
```
clock_t clock(void) __asm("_" "clock" "$UNIX2003");
```
:-(

**Comment by jacob-carlborg** [comment]
Looks like 3fa072618a8d9c45920ec6dea8a58dc3fe9535ad passed on all platforms except Windows where it, timed out?

**Comment by stefan-koch-sociomantic** [comment]
Does this still build with VS?

**Comment by stefan-koch-sociomantic** [comment]
@RazvanN7 I'd say this should not have a serious effect since it's in the optimizer which you should turn off if you want fast compiletime

File: src/dmd/backend/go.d [hunk 2]
```diff
@@ -0,0 +1,511 @@
...
+        //clock_t clock() { return 0; }
+    }
+}
+
+clock_t clock() { return 0; }
++/
+
+extern (C++):
+
+extern (C) clock_t os_clock();
```
Comment (rainers) [inline src/dmd/backend/go.d]: `os_clock` is extern(C++) despite the file extension of os.c.

**Comment by WalterBright** [comment]
Ready to go now.

File: src/dmd/backend/go.d [hunk 3]
```diff
@@ -0,0 +1,513 @@
+/**
+ * Compiler implementation of the
+ * $(LINK2 http://www.dlang.org, D programming language).
+ *
+ * Copyright:   Copyright (C) 1986-1998 by Symantec
+ *              Copyright (c) 2000-2017 by Digital Mars, All Rights Reserved
+ * Authors:     $(LINK2 http://www.digitalmars.com, Walter Bright)
+ * License:     Distributed under the Boost Software License, Version 1.0.
+ *              http://www.boost.org/LICENSE_1_0.txt
+ * Source:      https://github.com/dlang/dmd/blob/master/src/dmd/backend/go.c
```
Comment (rainers) [inline src/dmd/backend/go.d]: still links to C file.

File: src/dmd/backend/go.d [hunk 4]
```diff
@@ -0,0 +1,513 @@
...
+{
+    version (X86)
+    {
+        //pragma(mangle, "clock$UNIX2003") clock_t clock();
+        //clock_t clock() { return 0; }
+    }
+}
+
+clock_t clock() { return 0; }
++/
```
Comment (rainers) [inline src/dmd/backend/go.d]: test remainder?

File: src/dmd/backend/go.d [hunk 5]
```diff
@@ -0,0 +1,513 @@
...
+ */
+
+int go_flag(char *cp)
+{
+    enum GL     // indices of various flags in flagtab[]
+    {
+        O,all,cnp,cp,cse,da,dc,dv,li,liv,local,loop,
+        none,o,reg,space,speed,time,tree,vbe,MAX
+    }
+    __gshared const char*[GL.MAX] flagtab =
```
Comment (rainers) [inline src/dmd/backend/go.d]: Can you use `static immutable` here?

File: src/dmd/backend/go.d [hunk 6]
```diff
@@ -0,0 +1,513 @@
...
+    for (block* b = startblock; b; b = b.Bnext)
+    {
+        if (!b.Belem)
+            continue;
+        int d = el_countCommas(b.Belem) + 100;
+        if (d > iterationLimit)
+            iterationLimit = d;
+    }
+
+//    optfunc2(iterationLimit);
```
Comment (rainers) [inline src/dmd/backend/go.d]: Why is this comment added?

File: src/dmd/backend/global.d [hunk 7]
```diff
@@ -140,7 +140,8 @@ void util_set32();
 void util_set64();
 int ispow2(uint64_t);
 
-//#if __GNUC__
+version (Posix)
+{
```
Comment (rainers) [inline src/dmd/backend/global.d]: shouldn't the util_malloc/calloc/free functions use (templated) forwarding functions to not repeat the `version(Posix)` condition everywhere? Or just implement `util_*` in util2.c for Posix, too? The parc_* don't seem to be used anyway.

**Review by rainers — COMMENTED** [review]
Apart from nits, LGTM

File: src/dmd/backend/go.d (hunk 3 above)
Comment (WalterBright) [reply src/dmd/backend/go.d]: fixed

File: src/dmd/backend/go.d (hunk 4 above)
Comment (WalterBright) [reply src/dmd/backend/go.d]: removed

File: src/dmd/backend/go.d (hunk 6 above)
Comment (WalterBright) [reply src/dmd/backend/go.d]: removed

File: src/dmd/backend/global.d (hunk 7 above)
Comment (WalterBright) [reply src/dmd/backend/global.d]: The memory allocation functions in the back end are a bit of a mess. Straightening them out is something I plan to do, but it's beyond the scope of this PR.

File: src/dmd/backend/go.d (hunk 5 above)
Comment (WalterBright) [reply src/dmd/backend/go.d]: sure why not

**Comment by rainers** [comment]
It seems the OSX/32 failure is related to this PR. It does not happen always, but the ICE for std.conv can be found in earlier failures, too.

**Comment by WalterBright** [comment]
@rainers does that mean this problem is also happening with other than this PR?

**Comment by WalterBright** [comment]
It works fine on my OSX machine, sigh.

**Comment by WalterBright** [comment]
Adding instrumentation to the autotester, and now it works. What a surprise! A true heisenbug.

**Comment by rainers** [comment]
I bisected the (not so spurious) ICE, it is caused by a wrong definition of CLOCKS_PER_SEC in core.time. It should be 1_000_000 instead of 100 on the system I tested it on.

Not caused by the translation, but the ICE makes the optimization look brittle. Also, using a timeout is no good for reproducible builds, though a 30 second timeout for a single function is not going to be hit too often.

**Comment by WalterBright** [comment]
@rainers Thanks for figuring this out! Can you please submit a PR for core.stc.time? Also, what are you referring to with "the optimization"?

File: src/dmd/backend/go.d [hunk 8]
```diff
@@ -3,56 +3,129 @@
  * $(LINK2 http://www.dlang.org, D programming language).
  *
  * Copyright:   Copyright (C) 1986-1998 by Symantec
- *              Copyright (C) 2000-2018 by The D Language Foundation, All Rights Reserved
+ *              Copyright (c) 2000-2017 by Digital Mars, All Rights Reserved
```
Comment (wilzbach) [inline src/dmd/backend/go.d]: Is this change on purpose? Looks like an accidentally copied header from another file.

**Review by wilzbach — APPROVED** [review]
